File: frontend/src/utils/validation.ts
// Validation utilities for forms - updated

export interface ValidationResult {
  isValid: boolean;
  error?: string;
}

export interface PasswordStrength {
  score: number; // 0-4 (weak to very strong)
  feedback: string[];
  isValid: boolean;
}

// Email validation with comprehensive regex
export const validateEmail = (email: string): ValidationResult => {
  if (!email || email.trim() === '') {
    return { isValid: false, error: 'Email is required' };
  }

  // Comprehensive email regex that handles most valid email formats
  const emailRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;
  
  if (!emailRegex.test(email.trim())) {
    return { isValid: false, error: 'Please enter a valid email address' };
  }

  // Additional checks for common issues
  const trimmedEmail = email.trim().toLowerCase();
  
  // Check for consecutive dots
  if (trimmedEmail.includes('..')) {
    return { isValid: false, error: 'Email cannot contain consecutive dots' };
  }

  // Check for valid domain length
  const domain = trimmedEmail.split('@')[1];
  if (!domain || domain.length < 2) {
    return { isValid: false, error: 'Please enter a valid email domain' };
  }

  // Check for common typos in popular domains
  const commonDomainTypos = {
    'gmial.com': 'gmail.com',
    'gmai.com': 'gmail.com',
    'yahooo.com': 'yahoo.com',
    'hotmial.com': 'hotmail.com',
    'outlok.com': 'outlook.com'
  };

  if (commonDomainTypos[domain as keyof typeof commonDomainTypos]) {
    return { 
      isValid: false, 
      error: `Did you mean ${commonDomainTypos[domain as keyof typeof commonDomainTypos]}?` 
    };
  }

  return { isValid: true };
};

// Password strength validation
export const validatePasswordStrength = (password: string): PasswordStrength => {
  if (!password) {
    return {
      score: 0,
      feedback: ['Password is required'],
      isValid: false
    };
  }

  const feedback: string[] = [];
  let score = 0;

  // Length check (minimum 8 characters)
  if (password.length < 8) {
    feedback.push('Password must be at least 8 characters long');
  } else {
    score += 1;
    if (password.length >= 12) {
      score += 1; // Bonus for longer passwords
    }
  }

  // Character variety checks
  const hasLowerCase = /[a-z]/.test(password);
  const hasUpperCase = /[A-Z]/.test(password);
  const hasNumbers = /\d/.test(password);
  const hasSpecialChars = /[!@#$%^&*(),.?":{}|<>]/.test(password);

  if (!hasLowerCase) feedback.push('Add lowercase letters');
  if (!hasUpperCase) feedback.push('Add uppercase letters');
  if (!hasNumbers) feedback.push('Add numbers');
  if (!hasSpecialChars) feedback.push('Add special characters (!@#$%^&*)');

  // Score based on character variety
  const varietyCount = [hasLowerCase, hasUpperCase, hasNumbers, hasSpecialChars].filter(Boolean).length;
  score += varietyCount;

  // Common password checks
  const commonPasswords = [
    'password', '123456', '12345678', 'qwerty', 'abc123', 
    'password123', 'admin', 'letmein', 'welcome', '123456789'
  ];
  
  if (commonPasswords.includes(password.toLowerCase())) {
    feedback.push('Avoid common passwords');
    score = Math.max(0, score - 2);
  }

  // Sequential characters check
  const hasSequential = /(?:abc|bcd|cde|def|efg|fgh|ghi|hij|ijk|jkl|klm|lmn|mno|nop|opq|pqr|qrs|rst|stu|tuv|uvw|vwx|wxy|xyz|123|234|345|456|567|678|789)/i.test(password);
  if (hasSequential) {
    feedback.push('Avoid sequential characters');
    score = Math.max(0, score - 1);
  }

  // Repetitive characters check
  const hasRepetitive = /(.)\1{2,}/.test(password);
  if (hasRepetitive) {
    feedback.push('Avoid repetitive characters');
    score = Math.max(0, score - 1);
  }

  // Determine final score and validation
  const maxScore = 6; // length(2) + variety(4)
  score = Math.min(score, maxScore);
  
  const isValid = score >= 3 && password.length >= 8 && varietyCount >= 3;

  if (feedback.length === 0) {
    if (score >= 5) {
      feedback.push('Very strong password!');
    } else if (score >= 4) {
      feedback.push('Strong password');
    } else if (score >= 3) {
      feedback.push('Good password');
    }
  }

  return {
    score,
    feedback,
    isValid
  };
};

// Password confirmation validation
export const validatePasswordConfirmation = (password: string, confirmPassword: string): ValidationResult => {
  if (!confirmPassword) {
    return { isValid: false, error: 'Please confirm your password' };
  }

  if (password !== confirmPassword) {
    return { isValid: false, error: 'Passwords do not match' };
  }

  return { isValid: true };
};

// Get password strength color for UI
export const getPasswordStrengthColor = (score: number): string => {
  if (score <= 1) return '#e74c3c'; // Red
  if (score <= 2) return '#f39c12'; // Orange
  if (score <= 3) return '#f1c40f'; // Yellow
  if (score <= 4) return '#2ecc71'; // Green
  return '#27ae60'; // Dark green
};

// Get password strength label
export const getPasswordStrengthLabel = (score: number): string => {
  if (score <= 1) return 'Very Weak';
  if (score <= 2) return 'Weak';
  if (score <= 3) return 'Fair';
  if (score <= 4) return 'Strong';
  return 'Very Strong';
};

File: backend/validation.py
"""
Input validation utilities for the backend
"""
import re
from typing import List, Optional
from fastapi import HTTPException


def validate_email(email: str) -> tuple[bool, Optional[str]]:
    """
    Validate email format and return validation result
    
    Returns:
        (is_valid: bool, error_message: Optional[str])
    """
    if not email or not isinstance(email, str):
        return False, "Email is required"
    
    email = email.strip()
    
    if not email:
        return False, "Email cannot be empty"
    
    # Comprehensive email regex
    email_pattern = r'^[a-zA-Z0-9.!#$%&\'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$'
    
    if not re.match(email_pattern, email):
        return False, "Invalid email format"
    
    # Check email length (RFC 5321 limit)
    if len(email) > 254:
        return False, "Email address too long"
    
    # Check for consecutive dots
    if '..' in email:
        return False, "Email cannot contain consecutive dots"
    
    # Check domain part
    try:
        local, domain = email.rsplit('@', 1)
        if not domain or len(domain) < 2:
            return False, "Invalid email domain"
        
        # Check local part length (RFC 5321 limit)
        if len(local) > 64:
            return False, "Email local part too long"
            
    except ValueError:
        return False, "Invalid email format"
    
    return True, None


def validate_password_strength(password: str) -> tuple[bool, List[str]]:
    """
    Validate password strength requirements
    
    Returns:
        (is_valid: bool, error_messages: List[str])
    """
    if not password or not isinstance(password, str):
        return False, ["Password is required"]
    
    errors = []
    
    # Length check
    if len(password) < 8:
        errors.append("Password must be at least 8 characters long")
    
    # Character variety checks
    if not re.search(r'[a-z]', password):
        errors.append("Password must contain at least one lowercase letter")
    
    if not re.search(r'[A-Z]', password):
        errors.append("Password must contain at least one uppercase letter")
    
    if not re.search(r'\d', password):
        errors.append("Password must contain at least one number")
    
    if not re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
        errors.append("Password must contain at least one special character")
    
    # Check for common weak passwords
    weak_passwords = [
        'password', '123456', '12345678', 'qwerty', 'abc123',
        'password123', 'admin', 'letmein', 'welcome', '123456789',
        '1234567890', 'password1', 'qwertyuiop', 'asdfghjkl'
    ]
    
    if password.lower() in weak_passwords:
        errors.append("Password is too common and easily guessable")
    
    # Check for sequential characters
    if re.search(r'(abc|bcd|cde|def|efg|fgh|ghi|hij|ijk|jkl|klm|lmn|mno|nop|opq|pqr|qrs|rst|stu|tuv|uvw|vwx|wxy|xyz|123|234|345|456|567|678|789)', password.lower()):
        errors.append("Password should not contain sequential characters")
    
    # Check for repetitive characters
    if re.search(r'(.)\1{2,}', password):
        errors.append("Password should not contain repetitive characters")
    
    return len(errors) == 0, errors


def sanitize_input(input_str: str, max_length: Optional[int] = None) -> str:
    """
    Sanitize string input to prevent XSS and other injection attacks
    """
    if not isinstance(input_str, str):
        return ""
    
    # Remove null bytes and control characters
    sanitized = re.sub(r'[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]', '', input_str)
    
    # Strip whitespace
    sanitized = sanitized.strip()
    
    # Limit length if specified
    if max_length and len(sanitized) > max_length:
        sanitized = sanitized[:max_length]
    
    return sanitized


def validate_registration_input(email: str, password: str, agreed_to_terms: bool) -> dict:
    """
    Comprehensive validation for user registration
    
    Returns:
        dict with 'valid' boolean and 'errors' list
    """
    errors = []
    
    # Sanitize inputs
    email = sanitize_input(email, 254)  # RFC 5321 email length limit
    
    # Validate email
    email_valid, email_error = validate_email(email)
    if not email_valid:
        errors.append(email_error)
    
    # Validate password
    password_valid, password_errors = validate_password_strength(password)
    if not password_valid:
        errors.extend(password_errors)
    
    # Validate terms agreement
    if not agreed_to_terms:
        errors.append("You must agree to the Terms of Service and Privacy Policy")
    
    return {
        'valid': len(errors) == 0,
        'errors': errors,
        'sanitized_email': email.lower()  # Normalize email to lowercase
    }


def raise_validation_error(message: str, details: Optional[List[str]] = None):
    """
    Raise a properly formatted validation error
    """
    error_detail = {"message": message}
    if details:
        error_detail["validation_errors"] = details
    
    raise HTTPException(status_code=422, detail=error_detail)